File: flutter/startup/lib/services/transcript_manager.dart
import 'dart:convert';

enum TranscriptTypeX { user, agent }

enum TranscriptStatusX { inProgress, end, interrupted, unknown }

class TranscriptItem {
  final String id;
  final TranscriptTypeX type;
  final String text;
  final TranscriptStatusX status;
  const TranscriptItem({required this.id, required this.type, required this.text, required this.status});
}

class TranscriptManager {
  final List<TranscriptItem> items = [];

  bool upsert(TranscriptItem item) {
    final idx = items.indexWhere((e) => e.id == item.id && e.type == item.type);
    if (idx >= 0) {
      items[idx] = item;
    } else {
      items.add(item);
    }
    return true;
  }

  bool upsertFromJson(String json) {
    final parsed = _parseJson(json);
    if (parsed == null) return false;
    return upsert(parsed);
  }

  TranscriptItem? _parseJson(String json) {
    try {
      final obj = jsonDecode(json);
      if (obj is! Map<String, dynamic>) return null;
      final objType = (obj['object'] ?? '').toString();
      TranscriptTypeX type;
      if (objType == 'assistant.transcription') {
        type = TranscriptTypeX.agent;
      } else if (objType == 'user.transcription') {
        type = TranscriptTypeX.user;
      } else {
        return null;
      }
      final idAny = obj['turn_id'] ?? obj['message_id'] ?? DateTime.now().microsecondsSinceEpoch;
      final id = idAny.toString();
      final text = (obj['text'] ?? '').toString();
      final statusCode = _asInt(obj['turn_status']);
      TranscriptStatusX status;
      switch (statusCode) {
        case null:
          status = TranscriptStatusX.inProgress;
          break;
        case 0:
          status = TranscriptStatusX.inProgress;
          break;
        case 1:
          status = TranscriptStatusX.end;
          break;
        case 2:
          status = TranscriptStatusX.interrupted;
          break;
        default:
          status = TranscriptStatusX.unknown;
      }
      return TranscriptItem(id: id, type: type, text: text.isEmpty ? '(empty)' : text, status: status);
    } catch (_) {
      return null;
    }
  }

  int? _asInt(dynamic v) {
    if (v == null) return null;
    if (v is int) return v;
    if (v is String) return int.tryParse(v);
    return null;
  }
}

File: unity/startup/Assets/Scripts/Startup/TranscriptManager.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Startup
{
    public enum TranscriptType { User, Agent }
    public enum TranscriptStatus { InProgress, End, Interrupted, Unknown }

    [Serializable]
    public class TranscriptItem
    {
        public string Id;
        public TranscriptType Type;
        public string Text;
        public TranscriptStatus Status;
    }

    public class TranscriptManager
    {
        public readonly List<TranscriptItem> Items = new List<TranscriptItem>();

        public bool UpsertFromJson(string json)
        {
            try
            {
                var obj = new DictWrapper { json = json };
                var objectType = obj.GetString("object");
                TranscriptType type;
                if (objectType == "assistant.transcription") type = TranscriptType.Agent;
                else if (objectType == "user.transcription") type = TranscriptType.User;
                else return false;

                var id = (obj.GetInt("turn_id")?.ToString()) ?? obj.GetString("message_id") ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();
                var text = obj.GetString("text") ?? "(empty)";
                var statusCode = obj.GetInt("turn_status");
                var status = TranscriptStatus.Unknown;
                if (statusCode == null || statusCode == 0) status = TranscriptStatus.InProgress;
                else if (statusCode == 1) status = TranscriptStatus.End;
                else if (statusCode == 2) status = TranscriptStatus.Interrupted;

                var item = new TranscriptItem { Id = id, Type = type, Text = text, Status = status };
                var idx = Items.FindIndex(e => e.Id == id && e.Type == type);
                if (idx >= 0) Items[idx] = item; else Items.Add(item);
                return true;
            }
            catch
            {
                return false;
            }
        }

        [Serializable]
        private class DictWrapper
        {
            public string objectType;

            public string GetString(string key)
            {
                return ExtractString(key);
            }

            public int? GetInt(string key)
            {
                var s = ExtractString(key);
                if (int.TryParse(s, out var v)) return v;
                return null;
            }

            private string ExtractString(string key)
            {
                // JsonUtility not friendly with dynamic keys; fallback to naive parsing
                var idx = json.IndexOf("\"" + key + "\"");
                if (idx < 0) return null;
                var colon = json.IndexOf(':', idx);
                if (colon < 0) return null;
                var startQuote = json.IndexOf('"', colon + 1);
                var endQuote = json.IndexOf('"', startQuote + 1);
                if (startQuote < 0 || endQuote < 0) return null;
                return json.Substring(startQuote + 1, endQuote - startQuote - 1);
            }

            public string json;
        }
    }
}